third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp

```diff
@@ -489,6 +489,235 @@ static llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
 }
 #endif
 
+// Utility class aggregating information required to create a versioning
+// condition.
+class VersioningCondition {
+public:
+  VersioningCondition(Value S, Value BS) : S(S), BS(BS) {
+    assert(isValid() && "Invalid values supplied");
+  }
+  //  VersioningCondition() : S(), BS(){};
+  //  VersioningCondition(const VersioningCondition &other) {
+  //    S = other.S;
+  //    BS = other.BS;
+  //  }
+
+  // Create the condition: (S % BS == 0 && S > BS)
+  Value materialize(OpBuilder &builder, Location loc) const {
+    assert(S && BS && "Expecting valid values");
+    Value zero =
+        builder.createOrFold<arith::ConstantIntOp>(loc, 0, S.getType());
+    Value cmp1 = builder.create<arith::CmpIOp>(
+        loc, arith::CmpIPredicate::eq,
+        builder.create<arith::RemSIOp>(loc, S, BS), zero);
+    Value cmp2 =
+        builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::sgt, S, BS);
+    return builder.create<arith::AndIOp>(loc, cmp1, cmp2);
+  }
+
+private:
+  bool isValid() const {
+    Type SType = S.getType(), BSType = BS.getType();
+    if (!isa<IntegerType>(SType) || !isa<IntegerType>(BSType))
+      return false;
+
+    return cast<IntegerType>(SType).getWidth() ==
+           cast<IntegerType>(BSType).getWidth();
+  }
+
+  Value S;  // The length of a row/column.
+  Value BS; // The block size.
+};
+
+// Utility class responsible for collecting masked operation in a loop that are
+// amenable to having their mask dropped when the loop is versioned.
+class MaskedOpsCollector {
+  friend class LoopVersioner;
+
+public:
+  bool collectMaskedOps(scf::ForOp &forOp) {
+    // Nested loop aren't currently handled.
+    if (isa<scf::ForOp>(forOp->getParentOp()) ||
+        forOp->getParentOp()->template getParentOfType<scf::ForOp>())
```

Comment:
Why the first part of the condition is not covered by the second part of the condition?
```suggestion
    if (forOp->getParentOp()->template getParentOfType<scf::ForOp>())
```

Reply:
The 2nd part then condition checks whether the parent of this operation is a `scf::ForOp` so it would return true for this case:
```
scf.for(...) {
  scf.if (...) {
    scf.for  <== current loop
  }
}
```

Reply:
how about `if (forOp->template getParentOfType<scf::ForOp>())`?